lib/qbwc/request/vendors.rb

```diff
@@ -106,38 +165,106 @@ def update_xml_to_send(object, session_id)
-                    <Country>#{object['ship_from_address']['country'] if object['ship_from_address']}</Country>
-                  </ShipAddress>
-                  <Phone>#{object['vendor_address']['phone'] if object['vendor_address']}</Phone>
-                  <AltPhone>#{object['ship_from_address']['phone'] if object['ship_from_address']}</AltPhone>
-                  <Email>#{object['email']}</Email>
+                <ListID>#{object['list_id']}</ListID>
+                <EditSequence>#{object['edit_sequence']}</EditSequence>
+                <Name>#{object['name']}</Name>
+                <FirstName>#{object['firstname']}</FirstName>
+                <LastName>#{object['lastname']}</LastName>
+                <Phone>#{object['vendor_address']['phone'] if object['vendor_address']}</Phone>
+                <AltPhone>#{object['ship_from_address']['phone'] if object['ship_from_address']}</AltPhone>
+                <Email>#{object['email']}</Email>
+                #{add_fields(object, FIELD_MAP)}
+                #{sales_tax_country(object)}
+                #{reporting_period(object)}
+                #{add_refs(object)}
+                <VendorAddress>
+                  #{add_fields(object['vendor_address'], ADDRESS_MAP) if object['vendor_address']}
+                </VendorAddress>
+                <ShipAddress>
+                  #{add_fields(object['ship_from_address'], ADDRESS_MAP) if object['ship_from_address']}
+                </ShipAddress>
+                #{additional_contacts(object)}
+                #{additional_notes(object)}
+                #{contacts(object)}
               </VendorMod>
             </VendorModRq>
           XML
         end
 
         private
 
+        def add_refs(object)
+          fields = ""
+          REF_MAP.each do |qbe_name, flowlink_name|
+            full_name = object[flowlink_name] || config[flowlink_name]
+            fields += "<#{qbe_name}><FullName>#{full_name}</FullName></#{qbe_name}>" unless full_name.nil?
+          end
+
+          fields
+        end
+
+        def add_fields(object, mapping)
+          fields = ""
+          mapping.each do |qbe_name, flowlink_name|
+            fields += "<#{qbe_name}>#{object[flowlink_name]}</#{qbe_name}>\n" unless object[flowlink_name].nil?
+          end
+
+          fields
+        end
+
+        def sales_tax_country(object)
+          return "" unless SALES_TAX_COUNTRIES.include?(object['sales_tax_country'])
+          "<SalesTaxCountry>#{object['sales_tax_country']}</SalesTaxCountry>"
+        end
+
+        def reporting_period(object)
+          return "" unless REPORTING_PERIODS.include?(object['reporting_period'])
+          "<ReportingPeriod>#{object['reporting_period']}</ReportingPeriod>"
+        end
+
+        def additional_contacts(object)
+          return unless object['additional_contacts'] && object['additional_contacts'].is_a?(Array)
```

Comment:
To stay consistent, it might be good to return an empty string here and lines 244 and 255. I'm pretty sure they'll get turned into empty strings when used in the xml but it would be good just in case

Reply:
Good call. Not sure what would happen actually.